File: MTGSearch/src/main/java/com/dbottillo/mtgsearchfree/presenter/CardPresenterImpl.java
package com.dbottillo.mtgsearchfree.presenter;

import com.dbottillo.mtgsearchfree.interactors.CardsInteractor;
import com.dbottillo.mtgsearchfree.model.MTGCard;
import com.dbottillo.mtgsearchfree.util.Logger;
import com.dbottillo.mtgsearchfree.view.CardView;

import javax.inject.Inject;

public class CardPresenterImpl implements CardPresenter {

    private final CardsInteractor interactor;
    private CardView cardView;
    private final Logger logger;

    private final Runner<MTGCard> cardWrapper;

    @Inject
    public CardPresenterImpl(CardsInteractor interactor, Logger logger, RunnerFactory runnerFactory) {
        this.logger = logger;
        logger.d("created");
        this.interactor = interactor;
        this.cardWrapper = runnerFactory.simple();
    }

    @Override
    public void loadOtherSideCard(MTGCard card) {
        cardWrapper.run(interactor.loadOtherSideCard(card), new Runner.RxWrapperListener<MTGCard>() {
            @Override
            public void onNext(MTGCard card) {
                logger.d();
                cardView.otherSideCardLoaded(card);
            }

            @Override
            public void onError(Throwable e) {

            }

            @Override
            public void onCompleted() {

            }
        });
    }

    @Override
    public void init(CardView cardView) {
        this.cardView = cardView;
    }

}

File: MTGSearch/src/main/java/com/dbottillo/mtgsearchfree/view/views/CardPresenterImpl.kt
package com.dbottillo.mtgsearchfree.view.views

import com.dbottillo.mtgsearchfree.interactors.CardsInteractor
import com.dbottillo.mtgsearchfree.model.MTGCard
import com.dbottillo.mtgsearchfree.util.Logger
import com.dbottillo.mtgsearchfree.view.CardView

import javax.inject.Inject

class CardPresenterImpl @Inject constructor(private val interactor: CardsInteractor,
                                            private val logger: Logger) : CardPresenter {

    private lateinit var cardView: CardView

    init {
        logger.d("created")
    }

    override fun init(cardView: CardView) {
        this.cardView = cardView
    }

    override fun loadOtherSideCard(card: MTGCard) {
        interactor.loadOtherSideCard(card).subscribe{
            cardView.otherSideCardLoaded(it)
        }
    }

}